src/main/java/uk/nhs/hee/trainee/details/service/NtnGenerator.java

```diff
@@ -0,0 +1,303 @@
+      Curriculum curriculum = curriculaIterator.next();
+      String specialtyCode = curriculum.getCurriculumSpecialtyCode();
+
+      if (index > 0) {
+        if (curriculum.getCurriculumSubType().equals("SUB_SPECIALTY")) {
+          log.info("Appending sub-specialty '{}'.", specialtyCode);
+          sb.append(".");
+        } else {
+          log.info("Appending specialty '{}'.", specialtyCode);
+          sb.append("-");
+        }
+      } else {
+        log.info("Using '{}' as first specialty.", specialtyCode);
+      }
+
+      sb.append(specialtyCode);
+
+      if (index == 0 && Objects.equals(curriculum.getCurriculumName(), "AFT")) {
+        sb.append("-FND");
+        break;
+      }
+    }
+
+    log.info("Calculated specialty concat: '{}'.", sb);
+    return sb.toString();
+  }
+
+  /**
+   * Get the GMC/GDC reference number for the given personal details.
+   *
+   * @param personalDetails The personal details to get the reference number for.
+   * @return The GMC/GDC number, depending on which is valid.
+   */
+  private String getReferenceNumber(PersonalDetails personalDetails) {
+    String gmcNumber = personalDetails.getGmcNumber();
+    return gmcNumber.matches("\\d{7}") ? gmcNumber : personalDetails.getGdcNumber();
+  }
+
+  /**
+   * Get the NTN suffix for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to get the suffix for.
+   * @return The calculated suffix for the programme membership's NTN.
+   */
+  private String getSuffix(ProgrammeMembership programmeMembership) {
+    log.info("Calculating suffix.");
+    String trainingPathway = programmeMembership.getTrainingPathway();
+    log.info("Using training pathway '{}' to calculate suffix.", trainingPathway);
+
+    String suffix = switch (trainingPathway) {
+      case "CCT" -> "C";
+      case "CESR" -> "CP";
+      default -> {
+        List<Curriculum> curricula = programmeMembership.getCurricula();
+        List<Curriculum> sortedCurricula = filterAndSortCurricula(curricula);
+        String firstSpecialtyCode = sortedCurricula.get(0).getCurriculumSpecialtyCode();
+        log.info("Using specialty code '{}' to calculate suffix.", trainingPathway);
+
+        yield Objects.equals(firstSpecialtyCode, "ACA") ? "C" : "D";
+      }
+    };
+
+    log.info("Calculated suffix: '{}'.", suffix);
+    return suffix;
+  }
+
+  /**
+   * Filter the given curricula to only those that are current and sort them alphanumerically.
+   *
+   * @param curricula The curricula to filter and sort.
+   * @return The current curricula, sorted alphanumerically by subtype and code.
+   */
+  private List<Curriculum> filterAndSortCurricula(List<Curriculum> curricula) {
+    LocalDate now = LocalDate.now();
+
+    return curricula.stream()
+        .filter(c ->
+            c.getCurriculumSpecialtyCode() != null && !c.getCurriculumSpecialtyCode().isBlank())
+        .filter(c -> c.getCurriculumStartDate().isBefore(now))
```

Comment:
SQLserver 'between' is inclusive, so I think this might need to be `isBefore(now+1day)` or something more elegant, and something similar for `isAfter()`

Reply:
Yes, I've realised that while working on allowing future programmes to work. I'll include it in a separate PR though.